[clang-tidy] Extend `bugprone-sizeof-expression` with matching `P +- sizeof(T)` and `P +- N */ sizeof(T)` cases, add `cert-arr39-c` alias

zporky opened the pull request:
Improved `bugprone-sizeof-expression` check to find suspicious pointer arithmetic calculations where the pointer is offset by an `alignof()`, `offsetof()`, or `sizeof()` expression.

Pointer arithmetic expressions implicitly scale the offset added to or subtracted from the address by the size of the pointee type. Using an offset expression that is already scaled by the size of the underlying type effectively results in a squared offset, which is likely an invalid pointer that points beyond the end of the intended array.

```c
void printEveryEvenIndexElement(int *Array, size_t N) {
  int *P = Array;
  while (P <= Array + N * sizeof(int)) { // Suspicious pointer arithmetics using sizeof()!
    printf("%d ", *P);

    P += 2 * sizeof(int); // Suspicious pointer arithmetics using sizeof()!
  }
}
```

zporky committed:
Initial pointer arithmetics using sizeof matcher

zporky committed:
Working version of the checker.

zporky committed:
Code simplification with elevating string constants.

zporky committed:
filtering out char* like matches

zporky committed:
Filtering out the uninteresting types in imperative way.

zporky committed:
Reporting the suspicious type and its size.

zporky committed:
add initial test file

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SuspiciousPointerArithmeticsUsingSizeofCheck.cpp:
```diff
@@ -0,0 +1,84 @@
+//===--- SuspiciousPointerArithmeticsUsingSizeofCheck.cpp - clang-tidy --===//
```
Please make same length as closing comment.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SuspiciousPointerArithmeticsUsingSizeofCheck.cpp:
```diff
@@ -0,0 +1,84 @@
+//static const char *bin_op_bind = "ptr-sizeof-expression";	
+static constexpr llvm::StringLiteral BinOp{"bin-op"};
+static constexpr llvm::StringLiteral PointedType{"pointed-type"};
+static const auto IgnoredType = qualType(anyOf(asString("char"),asString("unsigned char"),asString("signed char"),asString("int8_t"),asString("uint8_t"),asString("std::byte"),asString("const char"),asString("const unsigned char"),asString("const signed char"),asString("const int8_t"),asString("const uint8_t"),asString("const std::byte")));
```
Please Clang-format.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SuspiciousPointerArithmeticsUsingSizeofCheck.cpp:
```diff
@@ -0,0 +1,84 @@
+
+void SuspiciousPointerArithmeticsUsingSizeofCheck::check(const MatchFinder::MatchResult &Result) {
+    const ASTContext &Ctx = *Result.Context;
+    const BinaryOperator* Matched = Result.Nodes.getNodeAs<BinaryOperator>(BinOp);
```
```suggestion
    const auto* Matched = Result.Nodes.getNodeAs<BinaryOperator>(BinOp);
```

Type is explicitly stated in statement. Same for nex line.

whisperity committed:
doc: Add documentation for the check

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SuspiciousPointerArithmeticsUsingSizeofCheck.h:
```diff
@@ -0,0 +1,30 @@
+//===--- SuspiciousPointerArithmeticsUsingSizeofCheck.h - clang-tidy -*- C++ -*-===//
```
Length.

whisperity replied:
This is not fixable without removing the `-*- C++ -*-` part, but that also should be there, because emacs users.

whisperity committed:
gsum: Synchronise contents by MERGE from upstream master

whisperity committed:
chore(doc): Added release notes entries

whisperity committed:
style: Format out line break from `CERTModule`

whisperity committed:
style: Apply some formatting fixes

whisperity committed:
test: [wip] Start rewriting the test file (+ tear out line ending whitespace)

whisperity committed:
test: Implemented the tests with proper format and `// CHECK` lines

whisperity committed:
style: [NFC] `clang-format` the check's main file

whisperity committed:
wip: Refurbish the checker, improve code quality

whisperity committed:
feat: Match `alignof()` and `offsetof()` as well

whisperity committed:
gsum: Synchronise contents by MERGE from upstream master

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -101,7 +101,7 @@ New checks
 - New :doc:`bugprone-suspicious-pointer-arithmetics-using-sizeof
   <clang-tidy/checks/bugprone/suspicious-pointer-arithmetics-using-sizeof>`
   check that finds suspicious pointer arithmetic calculations where the pointer
```
See previous release Release Notes for proper formatting.

whisperity committed:
style: `clang-format`

whisperity committed:
style: `clang-format`

whisperity committed:
style: Reformat RST changes

whisperity committed:
test: Add a test for the "runtime offset number * sizeof()" case

nicovank commented on clang-tools-extra/clang-tidy/bugprone/SuspiciousPointerArithmeticsUsingSizeofCheck.cpp:
```diff
@@ -0,0 +1,108 @@
+constexpr llvm::StringLiteral PointedType{"pointed-type"};
+constexpr llvm::StringLiteral ScaleExpr{"scale-expr"};
+
+const auto IgnoredTypes = qualType(
+    anyOf(asString("char"), asString("unsigned char"), asString("signed char"),
+          asString("int8_t"), asString("uint8_t"), asString("std::byte"),
+          asString("const char"), asString("const unsigned char"),
+          asString("const signed char"), asString("const int8_t"),
+          asString("const uint8_t"), asString("const std::byte")));
+const auto InterestingPointer = pointerType(unless(pointee(IgnoredTypes)));
+
```
```suggestion
```

This is unused.

nicovank reviewed (commented):
Add check and alias entries to [clang-tools-extra/docs/clang-tidy/checks/list.rst](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/docs/clang-tidy/checks/list.rst).

Check name: this is not only sizeof. Maybe `bugprone-suspicious-pointer-scaling-arithmetic`? Others may have other ideas.

The two examples in the check documentation would not be matched with the current version of this check as they mutiply the `sizeof` expression with a constant before addition. I feel like this is also a common pattern. Can this check also catch those?

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/suspicious-pointer-arithmetics-using-sizeof.rst:
```diff
@@ -0,0 +1,50 @@
+`ARR39-C. Do not add or subtract a scaled integer to a pointer
+<http://wiki.sei.cmu.edu/confluence/display/c/ARR39-C.+Do+not+add+or+subtract+a+scaled+integer+to+a+pointer>`_.
+
+Limitations
+-----------
+
+While incorrect from a technically rigorous point of view, the check does not
+warn for pointer arithmetics where the pointee type is ``char``
+(``sizeof(char) == 1``, by definition) on purpose.
```
When we evaluated the check on OSS projects, it was found to be noisy, yes. There were some cases where the ˋPtr + sizeof(*Ptr)ˋ, or ˋPtr + sizeof(T)ˋ expanded from a macro.

whisperity commented:
@nicovank 
> Add check and alias entries to [clang-tools-extra/docs/clang-tidy/checks/list.rst](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/docs/clang-tidy/checks/list.rst).
> 
> Check name: this is not only sizeof. Maybe `bugprone-suspicious-pointer-scaling-arithmetic`? Others may have other ideas.

The new check will be dropped and instead the functionality is going to be integrated into **ˋbugprone-sizeof-expressionˋ**, which already checks a lot more other cases and then some, related to pointer arithmetics as well.

The check's name would have been fine as-is now (and making it part of _ˋbugprone-sizeof-expressionˋ_ will be fine as well), because the vast majority of cases will have ˋsizeofˋ as the issue, and both ˋalignof()ˋ and ˋoffsetof()ˋ essentially scaled **with** ˋsizeofˋ internally, **AND** lets the user commit to the same bogus pattern by doing another scaling with the ˋ+ˋ/ˋ-ˋ operator.

@nicovank
> The two examples in the check documentation would not be matched with the current version of this check as they mutiply the `sizeof` expression with a constant before addition. I feel like this is also a common pattern. Can this check also catch those?

Yes, I have already added some deliberately failing tests for these last night, and I am engineering the appropriate matchers for this.

whisperity committed:
feat: Match `P + N * sizeof()`, move implementation to `bugprone-sizeof-expression`

whisperity committed:
doc: Document pointer arithmetic cases checked by `bugprone-sizeof-expression`

(Also document the other case which has been checked for a while but did not
have a documentation.)

NagyDonat commented on clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp:
```diff
@@ -285,6 +288,50 @@ void SizeofExpressionCheck::registerMatchers(MatchFinder *Finder) {
+  // This check does not warn on cases where the pointee type is "1 byte",
+  // as those cases can often come from generics and also do not constitute a
+  // problem because the size does not affect the scale used.
+  const auto PtrArithmeticIgnoredPointeeTypes = qualType(anyOf(
+      asString("char"), asString("unsigned char"), asString("signed char"),
+      asString("int8_t"), asString("uint8_t"), asString("std::byte"),
+      asString("const char"), asString("const unsigned char"),
+      asString("const signed char"), asString("const int8_t"),
+      asString("const uint8_t"), asString("const std::byte")));
```
Would it be possible to match the canonical type instead of the qualified one? It's a bit suspicious that you list all the `const` variants but e.g. don't list the `volatile` ones...

NagyDonat reviewed (approved):
LGTM, I'm satisfied with the overall quality.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SizeofExpressionCheck.cpp:
```diff
@@ -409,6 +456,43 @@ void SizeofExpressionCheck::check(const MatchFinder::MatchResult &Result) {
+    const auto *PointeeTy = Result.Nodes.getNodeAs<QualType>("pointee-type");
+    const auto *ScaleExpr =
+        Result.Nodes.getNodeAs<Expr>("sizeof-in-ptr-arithmetic-scale-expr");
+    const auto PointeeSize =
```
`auto` should not be used - type is not spelled explicitlty.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/sizeof-expression.rst:
```diff
@@ -164,6 +164,104 @@ hidden through macros.
+pointer arithmetic, as both return the "size" in bytes and not elements,
+potentially resulting in doubly-scaled offsets.
+
+
```
Excessive newline.

whisperity committed:
fix: Review comments, handle `template`, silence `P + sizeof(Buf)/sizeof(Buf[0])`-like

nicovank reviewed (approved):
Looks good to me 👍

whisperity committed:
gsum: Synchronise contents by MERGE from upstream master
